packages/api-fetch/src/middlewares/preloading.js

```diff
@@ -68,6 +68,21 @@ function createPreloadingMiddleware( preloadedData ) {
  * @return {Promise<any>} Promise with the response.
  */
 function prepareResponse( responseData, parse ) {
+	if (
+		typeof responseData.headers === 'object' &&
+		responseData.headers !== null
+	) {
+		Object.entries( responseData.headers ).forEach( ( [ key, value ] ) => {
+			if ( key.toLowerCase() === 'link' ) {
+				responseData.headers[ key ] = value.replace(
+					/<([^>]+)>/,
+					( /** @type {any} */ _, /** @type {string} */ url ) =>
+						`<${ encodeURI( url ) }>`
+				);
+			}
+		} );
+	}
```

Comment:
What does this have to do with preloading? Does it work when you disable preloading altogether?

Reply:
@ellatrix 

Commenting out [this line](https://github.com/WordPress/gutenberg/blob/d9f18e5561b6233f69d82fff65c9cf8488227572/packages/api-fetch/src/index.js#L193) in apiFetch does seem to resolve the issue, though it results in the expected console error.

I suspect the issue might be related to incorrect encoding being sent in the headers. Please let me know if you'd like me to test anything further or if there are specific scenarios I should focus on.

![Screenshot 2024-12-20 at 11 02 59 PM](https://github.com/user-attachments/assets/6e302f06-5dad-4b27-8b19-b269d238c542)

Thanks!

Reply:
Can you try making this empty instead? https://github.com/WordPress/gutenberg/blob/d9f18e5561b6233f69d82fff65c9cf8488227572/packages/api-fetch/src/middlewares/preloading.js#L11-L16

If that fixes it, then it's indeed related to preload I guess.

In that case, can you be a bit more descriptive on how the PR fixes the issue? What was the previous preloaded data and what is it after the fix? Are we adjusting the data at the right point? Why is the data fixed client side vs server side for example?

Reply:
yes, putting the empty cache object does resolve the issue.

The issue stems from the inability to send non-ASCII characters in HTTP headers when using the Response constructor.

This PR addresses the problem by encoding non-ASCII characters in header links before sending them. Without this encoding, an error is thrown. ( you have to add catch block in`getEntityRecord` function to check it )

```
TypeError: Response constructor: Cannot convert value in record<ByteString, ByteString> branch of (sequence<sequence<ByteString>> or record<ByteString, ByteString>) to ByteString because the character at index 23 has value 1608 which is greater than 255.
    prepareResponse http://localhost:8888/wp-content/plugins/gutenberg/build/api-fetch/index.min.js?ver=267f89098b708f8d02b1:680
    createPreloadingMiddleware http://localhost:8888/wp-content/plugins/gutenberg/build/api-fetch/index.min.js?ver=267f89098b708f8d02b1:637
    enhancedHandler http://localhost:8888/wp-content/plugins/gutenberg/build/api-fetch/index.min.js?ver=267f89098b708f8d02b1:178
```

### Problematic Header Link:

Before encoding:

```
<http://localhost:8888/ویدیو/hello-video-world>; rel="alternate"; type=text/html
```

After encoding

```
<http://localhost:8888/%D9%88%DB%8C%D8%AF%DB%8C%D9%88/hello-video-world>; rel="alternate"; type=text/html
```

The fix involves encoding the header link just before it is sent, It won't affect other headers and ASCII characters.

Since the issue is from client side handling of the header link, particularly in the Gutenberg api-fetch module, it seems appropriate to address it here.

let me know if anything else needs to be tested regarding this or any other module I should look into.

Thanks.